Return unique_ptr from buildTypedMesh

lindsayad opened the pull request:
With the most recent libmesh update in #15867, move constructors for
`MeshBase` derived objects are deleted. In #15901, I was relying on the compiler to
perform named return value optimization, which it is not required to do,
and in fact it did not do. This PR changes the `buildTypedMesh` API to
return a `std::unique_ptr<T>` because `std::unique_ptr` is innately
movable. Moreover per the good suggestion of @dschwen, I've added
protected methods in `MeshGenerator` that can be used by `MeshGenerator`
derived objects to easily build `MeshBase` objects without making calls
through their `MooseMesh _mesh` member.

lindsayad committed:
Return unique_ptr from buildTypedMesh

With the most recent libmesh update in #15867, move constructors for
`MeshBase` derived objects are deleted. In #15901, I was relying on the compiler to
perform named return value optimization, which it is not required to do,
and in fact it did not do. This PR changes the `buildTypedMesh` API to
return a `std::unique_ptr<T>` because `std::unique_ptr` is innately
movable. Moreover per the good suggestion of Daniel, I've added
protected methods in `MeshGenerator` that can be used by `MeshGenerator`
derived objects to easily build `MeshBase` objects without making calls
through their `MooseMesh _mesh` member.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/592359/) on 1bec17c wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/15919/site)

This comment will be updated on new commits.

lindsayad committed:
Create MooseMesh::determineUseDistributedMesh

This method is executed upon MooseMesh construction and whenever a user
calls MooseMesh::setParallelType. The addition of this method and its
execution in the MooseMesh constructor instead of in
MooseMesh::buildMeshBaseObject will be necessary because we will add
geometric relationship managers before the mesh base object is created
when we're not using mesh generators. MooseApp checks MooseMesh for
whether the mesh is distributed when geometric relationship managers are
added, we we have to determine that information before those geometric
rms are added

lindsayad committed:
Make sure that MooseMesh::isDistributedMesh() is accurate

dschwen commented on framework/src/meshgenerators/DistributedRectilinearMeshGenerator.C:
```diff
@@ -1040,13 +1040,14 @@ DistributedRectilinearMeshGenerator::buildCube(UnstructuredMesh & mesh,
-  auto mesh = _mesh->buildMeshBaseObject(MooseMesh::ParallelType::DISTRIBUTED);
+
+  // DistributedRectilinearMeshGenerator always generates a distributed mesh
+  _mesh->setParallelType(MooseMesh::ParallelType::DISTRIBUTED);
```
could the `setParallelType` call be moved into the wrapper?

lindsayad commented:
I did automate setParallelType in MooseMesh::buildTypedMesh. Looks like I just missed taking out this guy 

> On Oct 11, 2020, at 4:44 PM, Daniel Schwen <notifications@github.com> wrote:
> 
> ﻿
> @dschwen commented on this pull request.
> 
> In framework/src/meshgenerators/DistributedRectilinearMeshGenerator.C:
> 
> >    // We will set up boundaries accordingly. We do not want to call
>    // ghostGhostedBoundaries in which allgather_packed_range  is unscalable.
>    // ghostGhostedBoundaries will gather all boundaries to every single processor
>    _mesh->needGhostGhostedBoundaries(false);
> -  auto mesh = _mesh->buildMeshBaseObject(MooseMesh::ParallelType::DISTRIBUTED);
> +
> +  // DistributedRectilinearMeshGenerator always generates a distributed mesh
> +  _mesh->setParallelType(MooseMesh::ParallelType::DISTRIBUTED);
> could the setParallelType call be moved into the wrapper?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.

lindsayad committed:
Try speeding up nemesis repartitioning test

moosebuild commented:
Job [Valgrind](https://civet.inl.gov/job/593660/) on 3d634f2 : invalidated by @lindsayad

Try to get a client on which this has failed before (e.g. slow mobuild)
